### PR TITLE
Allow null parent to override unrecognized parents

### PR DESCRIPTION
**Pull request by PhantomWatson**
In other words, if both the root of a tree and disconnected branches are passed to Hash::nest(), return what connects to the root and ignore the disconnected branches.

Addresses https://github.com/cakephp/cakephp/issues/6127

### CODE REVIEW

**Comment by markstory** [comment]
How would someone use this new root option?

**Comment by PhantomWatson** [comment]
When retrieving all records from a tree-structured table (or any subset of records that include `parent_id = null`), a user would use this modified code to ensure that the values included in the result of `find('threaded')` are consistent, rather than some values being removed or retained depending on the order of the records returned by the database.

**Comment by markstory** [comment]
Ok, so if you were to make a test case, or example code on how to use this new feature, what would that code look like?

**Comment by PhantomWatson** [comment]
The use of `find('threaded')` or `Hash::nest()` would not change, just the inconsistent in/exclusion of records would. For example, if a database query returns the following records:

	array(
		'id' => 1,
		'parent_id' => null,
		'name' => 'Record 1'
	),
	array(
		'id' => 2,
		'parent_id' => null,
		'name' => 'Record 2'
	),
	array(
		'id' => 3,
		'parent_id' => 1,
		'name' => 'Record 1.1'
	),
	array(
		'id' => 4,
		'parent_id' => 3,
		'name' => 'Record 1.1.1'
	),
	array(
		'id' => 5,
		'parent_id' => 6,
		'name' => 'Record 3'
	)

With the defect still in play, a call to `find('threaded')` will either return

	array(
		0 => array(
			'ModelName' => array(
				'id' => 1,
				'parent_id' => null,
				'name' => 'Record 1'
			),
			'children' => array(
				0 => array( 
					'ModelName' => array(
						'id' => 3,
						'parent_id' => 1,
						'name' => 'Record 1.1'
					),
					'children' => array(
						0 => array(
							'ModelName' => array(
								'id' => 4,
								'parent_id' => 3,
								'name' => 'Record 1.1.1'
							),
							'children' => array()
						)
					)
				)
			)
		),
		1 => array(
			'ModelName' => array(
				'id' => 2,
				'parent_id' => null,
				'name' => 'Record 2'
			),
			'children' => array()
		)
	)

**or** it will return

	array(
		0 => array(
			'ModelName' => array(
				'id' => 5,
				'parent_id' => 6,
				'name' => 'Record 3'
			),
			'children' => array()
		)
	)

With the defect fixed, a call to `find('theaded')` will only return the first value.

If the docs for `find('threaded')` were to be updated to reflect this, I think that what would need to be added would be something along the lines of 
"If there are multiple possible roots of the tree (topmost parents), the output will place records with `parent_id = null` or `0` at the root if they are present, or otherwise the records with whichever topmost `parent_id` comes first in the results of the database query. Records which are not connected to the chosen root will not be included in the method's output."

File: lib/Cake/Test/Case/Model/ModelReadTest.php [hunk 1]
```diff
@@ -3578,29 +3578,28 @@ public function testFindAllThreaded() {
 		$expected = array(
 			array(
 				'Category' => array(
-					'id' => '2',
-					'parent_id' => '1',
-					'name' => 'Category 1.1'
+					'id' => '4',
+					'parent_id' => '0',
```
Comment (markstory) [inline lib/Cake/Test/Case/Model/ModelReadTest.php]: Odd that this isn't null.

**Comment by markstory** [comment]
The change in existing behavior makes me think that there is a possibility that this will result in other regressions being reported.

**Comment by ADmad** [comment]
Perhaps we can target this change to 2.7?  Always dropping off nodes for which parent cannot be found for in given dataset seems reasonable to me.

**Comment by lorenzo** [comment]
Just FYI, this is how the tree behavior in 3.0 works. A tree can have more than one root and they are all placed at the same level.

**Comment by PhantomWatson** [comment]
@ADmad, keep in mind that if nodes are *always* removed when their parents can't be found, then that would prevent ever getting an isolated branch of a tree, like just the apple branch from

    Vegetables
        Potatoes
    Fruits
        Bananas
        Apples
            Granny Smith
            Red Delicious
        Peaches

So the way that the code finds the first unrecognized parent and calls that the root makes perfect sense, but causes the probably-always-unintended side-effect of sometimes letting an unconnected node jack up an attempt to get the rest of the tree.

**Comment by dereuromark** [comment]
We should rebase this for 2.7, then it should be good to go, should it not?

**Comment by markstory** [comment]
@dereuromark Yeah it would have to go in 2.7 and be documented as something to watch out for when upgrading.

**Comment by PhantomWatson** [comment]
I merged in the latest 2.7 branch and issued a new pull request at https://github.com/cakephp/cakephp/pull/6491. It's a messy pull request with a bunch of @dereuromark and @markstory's commits bundled in, but I wasn't sure of a more elegant way to generate it.

I'm all ears if anyone has suggestions of a better way to have done that, though.

**Pull request closed without merging**